### DISCUSSION TITLE
TypeScript integration

### DISCUSSION THREAD

**Discussion by aboveyunhai** [codesandbox/sandpack#237] [Q&A]
I wonder is autocomplete avaliable in sandpack?
I see it used codemirror internally, but I did not find any apis about this feature. It's one of most important feature when we are looking for this kind of editors.

____
Edit: 
Under @danilowoz help, I was able to create a working Sandpack typescript in next.js

https://playground-ts.vercel.app/
Source Code: https://github.com/aboveyunhai/playground-ts

Something wrong with the initial render, but once you start to edit the code or change tab (trig render basically), ts-server will start to kick in.

There are some tiny optimization to avoid recreating ts-server for each instance. 
You might also be aware of some ref errors inside the code.

**Comment by danilowoz** [comment]
Hey @aboveyunhai, thanks for raising up this topic.
Curious to know which kind of autocomplete you mean, the [codemirror autosuggestion](https://codemirror.net/demo/complete.html)
or some kind of server language to autocomplete the reference in the document?

Let me know if you have any examples, and then we can consider them into our roadmap

**Reply by aboveyunhai** [reply]
> Hey @aboveyunhai, thanks for raising this topic here.
> Curious to know which kind of autocomplete you mean, the [codemirror autosuggestion](https://codemirror.net/demo/complete.html)
> or some kind of server language to autocomplete the reference in the document?
> 
> Let me know if you have any example, and then we can cons

I'm not exactly sure should I phase it as "autocomplete" or "hint" based on typescript declaration files. Monaco editor are able to create "hint" based on `d.ts`. 
https://playground-brainstorm-git-monaco-editor-react-4ed6d5-anubra266.vercel.app/ here is the link of the experiment I had built previously. It might be invalid in the future. There are two editors. One from sandpack and one is Monaco. So you get the idea when you type `onClick` inside `<Button />` component

**Reply by aboveyunhai** [reply]
To further expand the idea. Are we able to catch the declaration across all files from installed packages since some of them are `tsx` or `ts` files instead of `d.ts` though i believe they will all end up as `d.ts` file after compiling.

**Reply by aboveyunhai** [reply]
Codesandbox already had it but I'm not sure about sandpack?

**Reply by danilowoz** [reply]
Thanks for sharing your experiment! Sandpack uses Codemirror, we would need to implement by ourselves a language server to communicate with the editor. Unfortunately, we don't have plans in the short term to implement it due to other priorities, but I think we can produce a document of how you can support it out of the box.

Just found this information, which makes it possible, looking forward to giving it a try and making it work on Sandpack.

Thread: https://discuss.codemirror.net/t/codemirror-6-and-typescript-lsp/3398/22?page=2
Demo: https://okikio-codemirror.netlify.app/
Code: https://github.com/okikio/codemirror/blob/main/src/components/Editor.tsx

**Reply by aboveyunhai** [reply]
> 

will try it out later today, if the implementation is relatively simple without affecting with other functionalities. Maybe worth me to create a PR

**Comment by danilowoz** [comment]
Here's a working in progress sandbox:
https://codesandbox.io/s/github/danilowoz/sandpack-tsserver

**Reply by danilowoz** [reply]
Hey all! I'd like to share the latest status of the TypeScript LSP implementation. Short answer, it works! 🎉

You can find in this [link](https://codesandbox.io/s/github/danilowoz/sandpack-tsserver) the integration between TS Virtual File-System and Sandpack (mainly CodeMirror) and play with some examples. Fortunately, I could implement most of the features we could expect for a decent experience to code TypeScript, for example:
- IntelliSense;
- Tooltip error;
- Multiple files;
- Support tsconfig.json;
- Automatically dependency-types fetching (CodeSandbox CDN);
- In-browser dependency cache;

Demo: https://codesandbox.io/s/github/danilowoz/sandpack-tsserver

TBH I'm not an expert in LSP or IDE tools, so it's not clear to me what else this integration is missing, so please let me know if there's any feature request and/or if you'd like to collaborate in this project. I'm also going to post this progress in the CodeMirror Forum to get feedback and find out what's the potential problems with this implementation. 

**Sandpack challenges**

- How to distribute this implementation? `@codesandbox/sandpack-extensions`?
- Are we willing to officially support? Out of scope? 
- Is there any performance improvements to de done? I see more than one worker by Sandpack instance running.

**Reply by aboveyunhai** [reply]
Amazing! Thx for the hard work. I will try to integrate your work into the sample website later today to test how it looks like. And further check what we can learn from monaco editor.

**Reply by cliffordfajardo** [reply]
Nice this is awesome! 
> I could implement most of the features we could expect for a decent experience to code TypeScript

I played around with your linked codesandbox demo & I agree with your sentiment as well
I plan on using this typescript based version on a work project over the next weeks & I'm sure when I get down to the implementation - I'll pick up nuggests of feedback

----
Hey @chrtze - what do you think about this?
I started a discussion about it here: https://github.com/wbkd/react-flow/discussions/1789
Now with typescript integration, as shown above, this could perhaps enhance the DX/UX experience for `react-flow` 

React's using `sandpack` as well for their new docs: https://beta.reactjs.org/apis/usestate#examples-basic
Happy to help integrate/test this with `react-flow` docs    💪🤝

**Reply by okikio** [reply]
@danilowoz `SharedWorker` could be used instead of formal workers via [@okikio/sharedworker](https://github.com/okikio/sharedworker)
 (https://sharedworker.okikio.dev)

Also, I think support for vscode lsps is a better goal than just a Typescript version, I started working on it, but unfortunately I don't have time to complete the research. What I have thus far is in the [lsp-dev](https://github.com/okikio/codemirror/tree/) branch of [okikio/codemirror](https://github.com/okikio/codemirror), y'all could potentially finish what I started.

My project is based off of https://github.com/qualified/lsps, and https://github.com/FurqanSoftware/codemirror-languageserver, you could possibly use both projects to help complete lsps integration into Codemirror 6.

**Reply by jfmmm** [reply]
@danilowoz I tried getting this working on a CRA project I have but I get this error when it try to initialize `createVirtualTypeScriptEnvironment` in the worker.

```
Uncaught (in promise) Error: error TS2318: Cannot find global type 'Array'.
error TS2318: Cannot find global type 'Boolean'.
error TS2318: Cannot find global type 'CallableFunction'.
error TS2318: Cannot find global type 'Function'.
error TS2318: Cannot find global type 'IArguments'.
error TS2318: Cannot find global type 'NewableFunction'.
error TS2318: Cannot find global type 'Number'.
error TS2318: Cannot find global type 'Object'.
error TS2318: Cannot find global type 'RegExp'.
error TS2318: Cannot find global type 'String'.

    at createVirtualTypeScriptEnvironment (index.ts:58:11)
    at createTsSystem (tsserver.js:129:13)
```

Seem it's missing some libs. I tried copying the setup of this sandbox (tsconfig & package.json) as much as possible but nothing work. 

I eventually figure it might be that something extra is packaged with codesanbox that isn't there in this specific sandbox so I downloaded a zip file of this sandbox and tried running it locally instead of on codesandbox and sure enough I get the same exact error I have on my side.

Any idea what need to be added for this to work?

**Reply by danilowoz** [reply]
Hey @jfmmm, can you share the project? Then I can help debug it because, from this error message, I don't know what it might be.

**Reply by jfmmm** [reply]
Hi @danilowoz thanks for looking at this,

That won't be needed because as mentioned the issue is present on your example, [this one](https://codesandbox.io/s/github/danilowoz/sandpack-tsserver)

You can repro by NOT running it on codesandbox, basically just download the sandbox and run it locally.

I'm still not sure what thing on codesandbox make it work there and not in local, but I managed to solve the issue on my side.

The main issue was that the [default typescript template](https://github.com/codesandbox/sandpack/blob/main/sandpack-react/src/templates/vanilla-typescript.ts) of sandpack does not include a `target` in the `tsconfig` file and this lead to that error.

So what I did is just give my sandbox a new tsconfig that include one.

```
files['/tsconfig.json'] = `
{
  "compilerOptions": {
    "target": "${ts.ScriptTarget.ES2021}",
    "lib": [
      "dom",
      "ES2021"
    ],
    [...]
  },
  "include": [
    "src"
  ]
}
`;
```

Notice how I pass `ts.ScriptTarget.ES2021` instead of `es2021` as a string. It actually the value `8`. It's an enum.

I would expect to be able to pass `es2021` as a string, but there seem to be a bug in `typescript-vfs`.

if I pass a string, when it reach that [line](https://github.com/microsoft/TypeScript-Website/blob/v2/packages/typescript-vfs/src/index.ts#L168), It will get converted to 8 and then the following line try to run `toLowerCase()` on it and it throw since it's not a string.

But if I pass `8` instead it get converted to `es2021` and it work. So weird. 

I haven't reported that bug yet, fell free to do it if you have time. This is a side project at work that sadly I don't have a lot of time to work on right now :\

Just let me know so I don't double report it

Thanks!

**Reply by aboveyunhai** [reply]
In case if someone is interested, I will edit my question and leave the link of the semi-working next.js project based on the CodeSandBox example.

**Reply by danilowoz** [reply]
Interesting, because even running locally, I haven't faced this issue. Anyway, glad you managed it. 

If someone else is interested in running this project, you can do it by opening the following links: 
Editor: https://codesandbox.io/p/github/danilowoz/sandpack-tsserver
Preview: https://b460co-3000.preview.csb.app/

**Reply by jfmmm** [reply]
@danilowoz I think it might be a chrome only issue, I have it happen on chrome but not in firefox.

I also tried updating your demo to the latest version of `react-sandpack` and `codemirror` throw this error:
```
Config merge conflict for field override
    at combineConfig (https://6t1n25.csb.app/node_modules/
codemirror/state/dist/index.js:1424:212
[...]
```
https://codesandbox.io/s/stoic-tristan-6t1n25

Any idea what the issue could be?

**Reply by danilowoz** [reply]
I think I managed to fix it. Sandpack has a couple of breaking changes from v0.x to v1.x, and updating the API fixes it 
https://codesandbox.io/s/festive-hopper-l5wvv3?file=/src/App.tsx

If you're interested about taking a look, here's the migration guide https://sandpack.codesandbox.io/docs/releases/v1#migration-guide

**Reply by jfmmm** [reply]
@danilowoz Sadly I'm still getting the same error `Config merge conflict for field override` on both chrome and firefox

**Reply by aboveyunhai** [reply]
@jfmmm   you might want to check this one: https://codesandbox.io/s/adoring-bardeen-hzi6h7?file=/src/sandpack-components/CodeEditor.tsx, I remove most of them with only one example left since you don't know the differences between v0 and v1, I think less is better for you, then you can add samples from the broken example one by one by yourself.
The problem comes from the creation of extension via `codemirrorTypescriptExtensions()` while ts-server is not ready yet. add a useMemo() hook can "hide" the issue but it will fail the initial rendering on type check. This behavior is different than v0 @danilowoz , any ways to keep the process synchronized?

**Reply by peterbartels** [reply]
@aboveyunhai @jfmmm 
I forked the tsserver repo and made some improvements. 

Can you try to run this branch: https://github.com/peterbartels/sandpack-tsserver/tree/load-dependencies-async

I don't see the errors anymore. It loads slower because all the dependencies need to be loaded (that was actually the fix I made).

**Comment by cliffordfajardo** [comment]
If anyone else stumbles on this thread, found another good read on monaco vs codemirror technical blog post
1. https://blog.replit.com/code-editors
2. https://github.com/codesandbox/sandpack/issues/305

**Comment by jakeleboeuf** [comment]
[Vocs](https://vocs.dev/docs/guides/twoslash) is solving this elegantly with [Twoslash](https://www.typescriptlang.org/dev/twoslash/) 🔥

**Comment by 3xau1o** [comment]
codemirror [added LSP support last year ](https://discuss.codemirror.net/t/codemirror-lsp-client/9309), this is also an [example](https://codemirror-languageservice.js.org/typescript)
